### DISCUSSION TITLE
Prevent migrate on serve?

### DISCUSSION THREAD

**Discussion by benallfree** [pocketbase/pocketbase#4087] [Q&A]
Some PocketHost users are bricking their instances by uploading bad or duplicate migrations. Then, they can no longer access their admin to take corrective action.

Is there any flag to prevent migrations on serve and only run them explicitly with `migrate up`?

**Comment by ganigeorgiev** [comment]
No, there is no such flag at the moment. I'm not sure how useful it will be but I'll consider for the next release at least adding an option field to the `ServeConfig` struct.

Note that even on migrations failure the `pb_data` should remain unchanged because all unapplied migrations are executed in a transaction and committed only if everything is ok.

**Reply by benallfree** [reply]
`pocketbase serve` will exit upon migration failure. A broken migration prevents `pocketbase serve` from running.

**Reply by ganigeorgiev** [reply]
Yes, but the user application could still be broken if there is a missing system migration (eg. when updating binary versions) or when a collection/field is requested in the client-side code that is not inserted yet.

**Reply by ganigeorgiev** [reply]
I think the best way to handle it "safely" for now would be to recommend your users to create a `pb_migrations` copy (eg. `pb_migrations_old`) before uploading their new changes so that can easily revert in case of a failure.

Migrations failure are rare and usually could happen if you do separate changes in different environments and don't sync the migrations between them (eg. changing a field on prod and then later applying the same change manually on local which will create a migration with different timestamp and will attempt to reinsert the field as new one once redeployed).

**Reply by benallfree** [reply]
Yes, the approach I currently recommend to users is that they FTP into the live instance and remove the offending migrations.

This circumstance is surprisingly common. 

If I had a `pocketbase serve --no-migrate-up` flag, I would build a separate `migrate` button into the PocketHost dashboard where it would run `pocketbase migrate up` separately and report any failures. This would also be nice because it would limit downtime. Busy instances can stay running while migrations are applied.

**Reply by benallfree** [reply]
> Yes, but the user application could still be broken if there is a missing system migration (eg. when updating binary versions) 

That's a good point. System migrations seem good to always migrate at startup, but JS migrations are definitely proving to be problematic. Is there a way to differentiate? Like `--no-js-migrations` or something? Perhaps there is a useful distinction between system vs user migrations.

**Reply by ganigeorgiev** [reply]
> Is there a way to differentiate?

No, everything is added in one `migrate.MigrationsList`. Usually it shouldn't matter whether the migration comes from the binary or it is user defined one. I'm not sure, I'll have to think a little more on it.

**Reply by benallfree** [reply]
Small side note: I'm not sure how much emphasis you place on this for your releases, but I try to make all the PocketHost migrations and code backward compatible so if I need to roll back code, I don't have to worry about rolling back migrations. I eventually phase out old fields and associated code, but I almost always wait a week or two before cleaning up the old stuff. 

This becomes important for PocketHost users if they want to `downgrade` to an earlier version of PocketBase. If the downgrade crosses an incompatible migration boundary, they are going to have trouble. 

Just food for thought as you consider the other issue 👍
